scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,104 +11,402 @@ import scalafix.ScalafixConfig
+      *
+      * Unfortunately, meta selection chains are term refs while they could
+      * be type refs (e.g. `A` in `A.this.B` is a term). For this reason,
+      * we need to check both names to guarantee whether a name is in scope.
+      */
+    @inline
+    def lookupBothNames(name: String,
+                        in: g.Scope,
+                        disambiguatingOwner: Option[g.Symbol],
+                        disambiguatingNamespace: String): g.Symbol = {
+      val typeName = g.TypeName(name)
+      val typeNameLookup = in.lookup(typeName)
+      val symbol =
+        if (typeNameLookup.exists) typeNameLookup
+        else {
+          val termName = g.TermName(name)
+          val termNameLookup = in.lookup(termName)
+          if (termNameLookup.exists) termNameLookup
+          else g.NoSymbol
+        }
+
+      // Disambiguate overloaded symbols caused by name shadowing
+      if (symbol.isOverloaded) {
+        val alternatives = symbol.alternatives
+        disambiguatingOwner
+          .flatMap(o => alternatives.find(_.owner == o))
+          .getOrElse {
+            val substrings = alternatives.iterator
+              .filter(s => disambiguatingNamespace.indexOf(s.fullName) == 0)
+            // Last effort to disambiguate, pick sym with longest substring
+            if (substrings.isEmpty) alternatives.head
+            else substrings.maxBy(_.fullName.length)
+          }
+      } else symbol
+    }
+
+    /** Remove sequential prefixes from a concrete ref. */
+    def removePrefixes(ref: m.Ref, prefixes: List[m.Name]): m.Ref = {
+      /* Pattern match on `value`s of names b/c `stripThis` creates new names. */
+      def loop(ref: m.Term.Ref,
+               reversedPrefixes: List[m.Name]): List[m.Term.Ref] = {
+        reversedPrefixes match {
+          case prefix :: acc =>
+            val prefixValue = prefix.value
+            ref match {
+              case m.Term.Select(qual, name) =>
+                val qualAsRef = qual.asInstanceOf[m.Term.Ref]
+                if (name.value == prefixValue) loop(qualAsRef, acc)
+                else {
+                  // Make sure that removes names seq and reconstruct trees
+                  val nestedResult = loop(qualAsRef, reversedPrefixes)
+                  if (nestedResult.isEmpty) List(name)
+                  else List(m.Term.Select(nestedResult.head, name))
+                }
+              case name: m.Term.Name if name.value == prefixValue => Nil
+              case r => List(r)
+            }
+          case Nil => List(ref)
+        }
+      }
+
+      val transformedRef = ref.transform {
+        case m.Type.Select(qual, typeName) =>
+          val removed = loop(qual, prefixes.reverse)
+          if (removed.isEmpty) typeName
+          else m.Type.Select(removed.head, typeName)
+        case r => r
+      }
+
+      transformedRef.asInstanceOf[m.Ref]
+    }
+
+    /** Missing import and shortened name */
+    type Missing = (m.Ref, m.Ref)
```

Comment:
Seeing how this is used, below I would go for a plain case class and remove the `Hit` type alias

```scala
case class ShortenedName(name: m.Ref, toImport: Option[m.Ref])
```